### PR TITLE
8292302: Windows GetLastError value overwritten by ThreadLocalStorage::thread

### PR DESCRIPTION
**Pull request by kevinjwalls**
This is an MR which partially reverts JDK-8289091 such that JavaThread::threadObj() does not call Thread::current().

A JVMTI operation could call threadObj() and clear the Windows GetLastError value.

Partial, because I haven't reverted changes in JavaThread::print_on_error(), they aren't connected to the problems seen.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292302](https://bugs.openjdk.org/browse/JDK-8292302): Windows GetLastError value overwritten by ThreadLocalStorage::thread

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to [2cbaa186](https://git.openjdk.org/jdk/pull/10147/files/2cbaa186ea00622e27fbf1bcfe68760c7c1440f1)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [2cbaa186](https://git.openjdk.org/jdk/pull/10147/files/2cbaa186ea00622e27fbf1bcfe68760c7c1440f1)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to [8de54561](https://git.openjdk.org/jdk/pull/10147/files/8de54561dde3be5420897593cf1eaccb1aa1a69b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10147/head:pull/10147` \
`$ git checkout pull/10147`

Update a local copy of the PR: \
`$ git checkout pull/10147` \
`$ git pull https://git.openjdk.org/jdk pull/10147/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10147`

View PR using the GUI difftool: \
`$ git pr show -t 10147`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10147.diff">https://git.openjdk.org/jdk/pull/10147.diff</a>

</details>

### CODE REVIEW

**Review by plummercj — APPROVED** [review]
The new test looks good. I'll let someone with a better understanding of JDK-8289091 comment on the runtime changes.

File: src/hotspot/share/runtime/javaThread.cpp [hunk 1]
```diff
@@ -161,10 +161,8 @@ void JavaThread::set_threadOopHandles(oop p) {
 }
 
 oop JavaThread::threadObj() const {
-  Thread* current = Thread::current_or_null_safe();
-  assert(current != nullptr, "cannot be called by a detached thread");
-  guarantee(current != this || JavaThread::cast(current)->is_oop_safe(),
-            "current cannot touch oops after its GC barrier is detached.");
+  // Using Thread::current_or_null_safe() here risks that calling threadObj() can
+  // overwrite a native thread local, e.g. JVMTI operations clearing GetLastError on Windows.
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/javaThread.cpp:165]: Suggestion:
```
// Ideally we would verify the current thread is oop_safe when this is called, but as we can
// be called from a signal handler we would have to use Thread::current_or_null_safe(). That
// has overhead and also interacts poorly with GetLastError on Windows due to the use of TLS.
// Instead callers must verify oop safe access.
```

**Review by dholmes-ora — APPROVED** [review]
This is fine to me - modulo changing the comment as below. As I'm away on vacation I will pre-approve. Please be sure @dcubed-ojdk sees this before integrating.

Thanks.

File: src/hotspot/share/runtime/javaThread.cpp (hunk 1 above)
Comment (kevinjwalls) [reply src/hotspot/share/runtime/javaThread.cpp:165]: Thanks David, will update comment and check Dan is OK with this.

Comment (kevinjwalls) [reply src/hotspot/share/runtime/javaThread.cpp]: Updated comment.

File: test/jdk/com/sun/jdi/JdbLastErrorTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+import jdk.test.lib.process.OutputAnalyzer;
+import lib.jdb.JdbCommand;
+import lib.jdb.JdbTest;
+
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.Linker;
+import java.lang.foreign.SymbolLookup;
+import java.lang.invoke.MethodHandle;
+import java.lang.foreign.ValueLayout;
```
Comment (dcubed-ojdk) [inline test/jdk/com/sun/jdi/JdbLastErrorTest.java:42]: L41 and L42 should be swapped for proper alpha sort order.

File: test/jdk/com/sun/jdi/JdbLastErrorTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        System.loadLibrary("Kernel32");
+        SymbolLookup lookup = SymbolLookup.loaderLookup();
+        MethodHandle getLastError = linker.downcallHandle(
+            lookup.lookup("GetLastError").orElseThrow(),
+            FunctionDescriptor.of(ValueLayout.JAVA_INT));
+        MethodHandle setLastError = linker.downcallHandle(
+            lookup.lookup("SetLastError").orElseThrow(),
+            FunctionDescriptor.ofVoid(ValueLayout.JAVA_INT));
+
+        for (int i=0; i<10; i++) {
```
Comment (dcubed-ojdk) [inline test/jdk/com/sun/jdi/JdbLastErrorTest.java:60]: nit: should be space around '=' and '<'.

File: test/jdk/com/sun/jdi/JdbLastErrorTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            setLastError.invoke(42);
```
Comment (dcubed-ojdk) [inline test/jdk/com/sun/jdi/JdbLastErrorTest.java:61]: If you use final variable for the '42', then you can use the variable
here on L61 and below on L64. Then you can call is something cool:
```
static final int THE_ANSWER = 42;  // ... to life, the universe and everything!
```

**Review by dcubed-ojdk — APPROVED** [review]
I agree with your choices in your partial [BACKOUT] of:

[JDK-8289091](https://bugs.openjdk.org/browse/JDK-8289091) move oop safety check from SharedRuntime::get_java_tid() to JavaThread::threadObj()

Thanks for adding the new test. I like it. 

Concerning:
>  More follow up needed about the last error reset caused by GC. 
Have you filed a follow up bug in hotspot/gc for that issue?

Thumbs up on this fix and new test.

**Comment by kevinjwalls** [comment]
Thanks Chris, David and Dan.
Dan,  will update with those changes shortly, and now I know this one is going in I'll get that other stress test I had into a new JBS bug and link it.

**Pull request closed without merging**